Mathlib/Algebra/Category/ModuleCat/OverFilteredColimits.lean

```diff
@@ -0,0 +1,351 @@
+    erw [colimit.w_apply, Concrete.ι_repColimit_eq]
+  have eq₃ : r₁ * r₂ = colimit.ι ℛ j
+      (ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₁) *
+       ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₂)) := by
+    rw [map_mul]
+    erw [colimit.w_apply, colimit.w_apply, Concrete.ι_repColimit_eq, Concrete.ι_repColimit_eq]
+  rw [eq₃]
+  conv_rhs => rw [eq₁]; rhs; rw [eq₂]
+  rw [rep_smul, rep_smul, rep_smul, smulColimit.mul_smul]
+
+lemma smul_zero (r : colimit (C := ℜ𝔦𝔫𝔤) ℛ) : colimitSMulColimit (ℳ := ℳ) r 0 = 0 := by
+  rw [show r = colimit.ι ℛ (Concrete.indexRepColimit ℛ r) _ by
+    erw [Concrete.ι_repColimit_eq], rep_smul, smulColimit.smul_zero]
+
+lemma smul_add (r : colimit (C := ℜ𝔦𝔫𝔤) ℛ) (m₁ m₂ : colimit (C := 𝔄𝔟) ℳ) :
+    colimitSMulColimit r (m₁ + m₂) = colimitSMulColimit r m₁ + colimitSMulColimit r m₂ := by
+  rw [show r = colimit.ι ℛ (Concrete.indexRepColimit ℛ r) _ by
+    erw [Concrete.ι_repColimit_eq], rep_smul, rep_smul, rep_smul, smulColimit.smul_add]
+
+lemma add_smul (r₁ r₂ : colimit (C := ℜ𝔦𝔫𝔤) ℛ) (m : colimit (C := 𝔄𝔟) ℳ) :
+    colimitSMulColimit (r₁ + r₂) m = colimitSMulColimit r₁ m + colimitSMulColimit r₂ m := by
+  classical
+  let O : Finset J :=
+    {  Concrete.indexRepColimit ℛ r₁, Concrete.indexRepColimit ℛ r₂ }
+  let H : Finset ((X : J) ×' (Y : J) ×' (_ : X ∈ O) ×' (_ : Y ∈ O) ×' (X ⟶ Y)) := {}
+  let j : J := IsFiltered.sup O H
+  have eq₁ : r₁ = colimit.ι ℛ j
+      (ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₁)) := by
+    erw [colimit.w_apply, Concrete.ι_repColimit_eq]
+  have eq₂ : r₂ = colimit.ι ℛ j
+      (ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₂)) := by
+    erw [colimit.w_apply, Concrete.ι_repColimit_eq]
+  have eq₃ : r₁ + r₂ = colimit.ι ℛ j
+      (ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₁) +
+       ℛ.map (IsFiltered.toSup O H $ by simp [O]) (Concrete.repColimit ℛ r₂)) := by
+    rw [map_add]
+    erw [colimit.w_apply, colimit.w_apply, Concrete.ι_repColimit_eq, Concrete.ι_repColimit_eq]
+  rw [eq₃]
+  conv_rhs => rw [eq₁]; rhs; rw [eq₂]
+  rw [rep_smul, rep_smul, rep_smul, smulColimit.add_smul]
+
+lemma zero_smul (m : colimit (C := 𝔄𝔟) ℳ) :
+    colimitSMulColimit (0 : colimit (C := ℜ𝔦𝔫𝔤) ℛ) m = 0 := by
+  let c : J := (inferInstance : IsFiltered J).2.some
+  rw [show (0 : colimit (C := ℜ𝔦𝔫𝔤) ℛ) = colimit.ι ℛ c 0 by rw [map_zero], rep_smul,
+    smulColimit.zero_smul]
+
+end colimitSMulColimit
+
+noncomputable instance moduleColimitColimit :
+    Module (colimit (C := ℜ𝔦𝔫𝔤) ℛ) (colimit (C := 𝔄𝔟) ℳ) where
+  smul := colimitSMulColimit
+  one_smul := colimitSMulColimit.one_smul _ _
+  mul_smul := colimitSMulColimit.mul_smul _ _
+  smul_zero := colimitSMulColimit.smul_zero _ _
+  smul_add := colimitSMulColimit.smul_add _ _
+  add_smul := colimitSMulColimit.add_smul _ _
+  zero_smul := colimitSMulColimit.zero_smul _ _
+
+lemma smul_spec
```

Comment:
Why do you define `r` and `m` in this lemma, and then have hypotheses that they're equal to other things? Is the lemma more useful in this form, or can you just delete `r` and `m` completely, and replace them with their values according to h1 and h2 (which you can also remove)?

Reply:
Thanks, copilot wrote this lemma actually xd